### PR TITLE
Add support for using a shell to run commands

### PR DESCRIPTION
**Pull request by alexkappa**
This PR adds support for supplying a shell to use when running commands. This is useful with expanding wildcards among other things.

Fixes: #229

### CODE REVIEW

File: packages/exec/src/interfaces.ts [hunk 1]
```diff
@@ -12,6 +12,9 @@ export interface ExecOptions {
   /** optional.  defaults to false */
   silent?: boolean
 
+  /** optional. Runs command inside of a shell. A different shell can be specified as a string. Defaults to false */
+  shell?: boolean | string
```
Comment (ericsciple) [inline packages/exec/src/interfaces.ts]: How does this affect the quoting rules on windows? linux?

Lots of logic in tool runner to accurately deal with arg quoting for .exe and different rules for .cmd/.bat.

Today not a problem on Linux because iirc node calls execv which takes an arg array. Different on Windows because everything gets joined into a single string. Does this introduce arg quoting challenges on Linux?

Comment (ericsciple) [reply packages/exec/src/interfaces.ts]: And if so, are the quoting rules different depending on the shell?

Comment (alexkappa) [reply packages/exec/src/interfaces.ts]: Hi @ericsciple, from the documentation of this options:

> - `shell` `<boolean>` | `<string>` If true, runs command inside of a shell. Uses '/bin/sh' on Unix, and process.env.ComSpec on Windows. A different shell can be specified as a string. See Shell Requirements and Default Windows Shell. Default: false (no shell).
> - `windowsVerbatimArguments` `<boolean>` No quoting or escaping of arguments is done on Windows. Ignored on Unix. This is set to true automatically when shell is specified and is CMD. Default: false.

Comment (alexkappa) [reply packages/exec/src/interfaces.ts]: Perhaps on windows we could set `windowsVerbatimArguments` to `true` if `shell` is specified and is equal to 'CMD'. That way we match the behavior of the stdlib, and allow our quoting to work with that assumption.

On linux I don't know if there are any quoting differences, but I'd be happy to add more tests to find out.

Comment (ericsciple) [reply packages/exec/src/interfaces.ts]: Thanks. Other thoughts, spitballing...

If drastically different, I wonder whether it would make more sense to be linux/mac only feature.

If quoting issues on Linux, then existing workaround of `tool: 'sh', args: ['-c', 'ls -l *.md']` is at least transparent what is happening

Comment (alexkappa) [reply packages/exec/src/interfaces.ts]: I am not entirely sure how to proceed with this @ericsciple. I understand the workaround, but I thought since node supports it natively, it could be low hanging fruit with a more elegant solution.

Will more tests alleviate your concerns? I understand you are reluctant to accept this as there might be edge cases that break quoting. If so, I might need some help :)

File: packages/exec/__tests__/exec.test.ts [hunk 2]
```diff
@@ -153,6 +153,23 @@ describe('@actions/exec', () => {
...
   })
 
+  it('Runs exec successfully with shell wildcard substitution', async () => {
+    let tool: string
+    let args: string[]
+    let opts: im.ExecOptions
+    if (IS_WINDOWS) {
+      tool = 'cmd'
+      args = ['/c', 'echo', 'hello']
+      opts = {shell: process.env.ComSpec}
```
Comment (craigdbarber) [inline packages/exec/__tests__/exec.test.ts]: Suggest also adding a test for running bash shell within Windows. As bash is now supported by windows runners, it provides a universal target for actions wanting to execute shell commands.

**Comment by imbsky** [comment]
What is the status of this PR? This change is required for `actions/cache`.

**Review by Lelee703 — APPROVED** [review]

File: packages/exec/__tests__/exec.test.ts (hunk 2 above)
Comment (Lelee703) [reply packages/exec/__tests__/exec.test.ts:163]: ok

**Review by Lelee703 — APPROVED** [review]

File: packages/exec/src/interfaces.ts (hunk 1 above)
Comment (Lelee703) [reply packages/exec/src/interfaces.ts:16]: same here

**Review by Lelee703 — APPROVED** [review]

**Comment by timharris777** [comment]
Why hasn't this been supported yet? I see that `@actions/exec` is actually using `child_process.spawn` under the hood. I agree with @alexkappa that adding  the `shell` option as an available option to pipe to the spawn command should be quite easy. I will be forking @alexkappa's fork as I have use of this.
